---
title: "Tutorial for PALM"
author: 
- name: "Zhoujingpeng Wei"
  affiliation: 
  - Department of Biostatistics and Medical Informatics, University of Wisconsin-Madison
  email: zwei74@wisc.edu
package: PALM
date: "`r Sys.Date()`"
output: 
  BiocStyle::html_document:
    toc: true
    toc_depth: 2
    toc_float: true
    code_folding: show
vignette: >
  %\VignetteIndexEntry{PALM_vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

# Introduction

This R package implements `PALM` [1], a quasi-Poisson-based framework designed for robust, scalable, and reproducible identification of covariate-associated microbial features in large-scale microbiome association studies and meta-analyses. The package can perform single-study association analysis and meta-analysis across multiple studies. This tutorial provides examples demonstrating meta-analysis.

```{r, echo=FALSE, out.width="80%"}
knitr::include_graphics("./Intro.pdf")
```

# Installation

Install package from GitHub.

```{r getPackage, echo=TRUE}
if(!require("PALM", quietly = TRUE)){
  devtools::install_github("ZjpWei/PALM_package")
}
```

Load the required packages. 

```{r load, echo=TRUE, message=FALSE, warning=FALSE}
library("PALM")
library("ggplot2")
library("cowplot")
library("tidyverse")
library("DT")
options(DT.options = list(
  initComplete = JS("function(settings, json) {",
  "$(this.api().table().header()).css({'background-color': 
  '#000', 'color': '#fff'});","}")))
```

# Meta-analysis given all samples of all participating studies

There are two approaches PALM can run meta-analysis. 

Approach #1: If the individual-level sample data across all participating studies are available, the function `palm` can take these data as input to generate, harmonize, and combine summary association statistics across studies for fast and reliable identification of microbial features. 

Approach #2: If individual-level samples of each study are not in a central place, summary statistics can be generated for each study separately using the functions `palm.null.model` and `palm.get.summary`. The summary statistics can be transported to a central place to be harmonized and combined for feature selection using the function `palm.meta.summary`.

We demonstrate Approach #1 in this session and Approach #2 in the next session.

Here we use the datasets from two metagenomics studies of colorectal cancer (CRC) [2] to demonstrate the use of each function. The `CRC_abd` is a list of sample-by-feature matrices of relative abundance counts of 267 species under order *Clostridiales* from the two studies. The `CRC_meta` is a data frame including the sample-level variables from these two studies. In particular, the following variables are in the `CRC_meta` data:

* Sample identity: "Sample_ID"

* Study name: "Study"

* Disease status: "Group"

```{r echo=TRUE}
## Load CRC data
data("CRC_data")
CRC_abd <- CRC_data$CRC_abd
CRC_meta <- CRC_data$CRC_meta
```


```{r echo=TRUE, message=TRUE, warning=FALSE}
# Prepare input data
rel.abd <- list()
covariate.interest <- list()
for(d in unique(CRC_meta$Study)){
  rel.abd[[d]] <- CRC_abd[CRC_meta$Sample_ID[CRC_meta$Study == d],]
  disease <- as.numeric(CRC_meta$Group[CRC_meta$Study == d] == "CRC")
  names(disease) <- CRC_meta$Sample_ID[CRC_meta$Study == d]
  covariate.interest[[d]] <- matrix(disease, ncol = 1, dimnames = list(names(disease), "disease"))
}

# palm analysis
meta.result <- palm(rel.abd = rel.abd, covariate.interest = covariate.interest)
```

The table below presents 20 microbial features, including their coefficient estimates, standard errors, coefficient significance p-values, and adjusted FDRs, as well as heterogeneity significance p-values and adjusted FDRs from the meta-analysis. Additionally, it includes the summary statistics estimates and standard errors for each individual study.

```{r}
meta.result$disease %>% dplyr::slice(1:20) %>% 
  mutate(across(where(is.numeric), ~ round(.x, digits = 4))) %>% 
  datatable()
```

The figure below displays the features with adjusted coefficient $FDR â‰¤ 0.05$. Each dot in scatter plot represents the coefficient estimate, with error bars indicating the standard errors. The bar plot shows the $-\log_{10}(\text{FDR})$ values.

```{r}
## summarize results
palm.df <- meta.result$disease %>% dplyr::filter(qval <= 0.05) %>%
  dplyr::filter(coef > 0) %>% arrange(qval) %>%
  dplyr::add_row(meta.result$disease %>% dplyr::filter(qval <= 0.05)) %>%
  dplyr::filter(coef < 0) %>% arrange(desc(qval)) %>%
  dplyr::transmute(feature = factor(feature, levels = feature), coef, stderr, qval, sig = (qval <= 0.05),
                   dir = case_when(coef > 0 & qval <= 0.05 ~ "pos",
                                   coef < 0 & qval <= 0.05 ~ "neg",
                                   TRUE ~ "other")) %>% arrange(desc(row_number()))  

## generate bar plot
df.bar <- palm.df %>% ggplot(aes(x=feature, y=-log10(qval), fill= sig)) +
      geom_bar(stat="identity") + ylim(0, 3) + scale_fill_manual(
        breaks = c(TRUE, FALSE), values = c("grey50", "grey80")) +
      geom_hline(aes(yintercept = -log10(0.05)),colour="#990000", linetype="dashed") +
      theme_minimal() + coord_flip() +
         theme(axis.title.x = element_blank(),
               axis.title.y = element_blank(),
               axis.ticks = element_blank(),
               panel.grid = element_blank(),
               plot.title = element_text(hjust = 0.5, size = 20),
               panel.background = element_rect(fill=NULL, colour='black', linewidth = 1),
               axis.text.y = element_blank(),
               axis.text.x =  element_text(size = 10),
               legend.title = element_text(hjust = 0.5, size = 16),
               legend.text = element_text(size = 10),
               legend.position = "none",
               legend.direction = "vertical",
               legend.box = "vertical",
               strip.text = element_blank())

## generate scatter plot
df.scatter <- palm.df %>% ggplot(aes(x=feature, y=coef)) +
      geom_errorbar(aes(ymin = coef - stderr, ymax = coef + stderr), color = "grey80",
                    width = 0, linewidth = 0.5) +
      geom_point(aes(color = dir), pch = 18, size = 2.5) + ylim(-2, 0.5) +
      geom_hline(aes(yintercept = 0),colour="#990000", linetype="dashed") +
      theme_minimal() + scale_color_manual(
        breaks = c("pos", "other", "neg"),
        values = c("red", "grey80", "blue")) +
      coord_flip() +
  theme(axis.title.x = element_blank(),
        axis.title.y = element_blank(),
        axis.ticks = element_blank(),
        panel.grid = element_blank(),
        plot.title = element_text(hjust = 0.5, size = 20),
        panel.background = element_rect(fill=NULL, colour='black', linewidth = 1),
        axis.text.y = element_text(size = 10),
        axis.text.x =  element_text(size = 10),
        legend.title = element_text(hjust = 0.5, size = 16),
        legend.text = element_text(size = 10),
        legend.position = "none",
        legend.direction = "vertical",
        legend.box = "vertical",
        strip.text = element_blank())

## plot
plot_grid(df.scatter, df.bar, nrow = 1, align = 'h',  rel_widths = c(0.72, 0.28))
```

# Meta-analysis given samples of one study at a time
If the datasets of different studies live in different locations and cannot be conveniently shared among studies, we can first generate summary statistics for each study:

```{r echo=TRUE, message=FALSE, warning=FALSE}
## Generate summary statistics for each study
null.obj.FR <- palm.null.model(rel.abd = rel.abd$`FR-CRC`)
summary.stats.FR <- palm.get.summary(null.obj = null.obj.FR, covariate.interest = covariate.interest$`FR-CRC`)

null.obj.DE <- palm.null.model(rel.abd = rel.abd$`DE-CRC`)
summary.stats.DE <- palm.get.summary(null.obj = null.obj.DE, covariate.interest = covariate.interest$`DE-CRC`)
```

These summary statistics can be transported to a central location for meta-analysis:

```{r echo=TRUE, message=FALSE, warning=FALSE}
## Concatenate summary statistics
summary.stats.merge <- c(summary.stats.FR, summary.stats.DE)
names(summary.stats.merge) <- c("FR-CRC", "DE-CRC")

## Meta-analysis to harmonize and combine summary statistics across studies
meta.result.2 <- palm.meta.summary(summary.stats = summary.stats.merge)
```

The meta-analysis results generated in this way are identical to those from the previous session. The following scatter plot and bar plot appear identical to those in the previous section.

```{r}
## summarize results
palm.df <- meta.result.2$disease %>% dplyr::filter(qval <= 0.05) %>%
  dplyr::filter(coef > 0) %>% arrange(qval) %>%
  dplyr::add_row(meta.result.2$disease %>% dplyr::filter(qval <= 0.05)) %>%
  dplyr::filter(coef < 0) %>% arrange(desc(qval)) %>%
  dplyr::transmute(feature = factor(feature, levels = feature), coef, stderr, qval, sig = (qval <= 0.05),
                   dir = case_when(coef > 0 & qval <= 0.05 ~ "pos",
                                   coef < 0 & qval <= 0.05 ~ "neg",
                                   TRUE ~ "other")) %>% arrange(desc(row_number()))  

## generate bar plot
df.bar <- palm.df %>% ggplot(aes(x=feature, y=-log10(qval), fill= sig)) +
      geom_bar(stat="identity") + ylim(0, 3) + scale_fill_manual(
        breaks = c(TRUE, FALSE), values = c("grey50", "grey80")) +
      geom_hline(aes(yintercept = -log10(0.05)),colour="#990000", linetype="dashed") +
      theme_minimal() + coord_flip() +
         theme(axis.title.x = element_blank(),
               axis.title.y = element_blank(),
               axis.ticks = element_blank(),
               panel.grid = element_blank(),
               plot.title = element_text(hjust = 0.5, size = 20),
               panel.background = element_rect(fill=NULL, colour='black', linewidth = 1),
               axis.text.y = element_blank(),
               axis.text.x =  element_text(size = 10),
               legend.title = element_text(hjust = 0.5, size = 16),
               legend.text = element_text(size = 10),
               legend.position = "none",
               legend.direction = "vertical",
               legend.box = "vertical",
               strip.text = element_blank())

## generate scatter plot
df.scatter <- palm.df %>% ggplot(aes(x=feature, y=coef)) +
      geom_errorbar(aes(ymin = coef - stderr, ymax = coef + stderr), color = "grey80",
                    width = 0, linewidth = 0.5) +
      geom_point(aes(color = dir), pch = 18, size = 2.5) + ylim(-2, 0.5) +
      geom_hline(aes(yintercept = 0),colour="#990000", linetype="dashed") +
      theme_minimal() + scale_color_manual(
        breaks = c("pos", "other", "neg"),
        values = c("red", "grey80", "blue")) +
      coord_flip() +
  theme(axis.title.x = element_blank(),
        axis.title.y = element_blank(),
        axis.ticks = element_blank(),
        panel.grid = element_blank(),
        plot.title = element_text(hjust = 0.5, size = 20),
        panel.background = element_rect(fill=NULL, colour='black', linewidth = 1),
        axis.text.y = element_text(size = 10),
        axis.text.x =  element_text(size = 10),
        legend.title = element_text(hjust = 0.5, size = 16),
        legend.text = element_text(size = 10),
        legend.position = "none",
        legend.direction = "vertical",
        legend.box = "vertical",
        strip.text = element_blank())

## plot
plot_grid(df.scatter, df.bar, nrow = 1, align = 'h',  rel_widths = c(0.72, 0.28))
```

# Meta-analysis of association scan for multiple covariates of interest
Microbiome association studies can involve a large number of covariates of interest (e.g., omics variables). We show here how to  use PALM to meta-analyze eight microbiome-metabolome association studies [3]. In these eight studies, we have 101 genera and 450 metabolites and we are interested in identifying genera that are associated with individual metabolites. This analysis takes approximately 20 minutes.

```{r echo=TRUE, message=FALSE, warning=FALSE}
## Load metabolite data: 
## You may see the following website on how to directly load data from 
## github into R https://github.com/ZjpWei/Melody/raw/main/Metabolite.rda
load(file = url("https://github.com/ZjpWei/Melody/raw/main/Metabolite.rda"))

## Change genera names
for(d in names(otu_data_lst)){
  colnames(otu_data_lst[[d]]) <-  gsub(".*;g__", "", colnames(otu_data_lst[[d]]))
}

## Get null model
null.obj <- palm.null.model(rel.abd = otu_data_lst, covariate.adjust = covariates_adjust_lst)

## Get summary statistics
summary.stat <- palm.get.summary(null.obj = null.obj, covariate.interest = cmpd_data_lst,
                                 cluster = cluster_data_lst)

## Meta-analysis
meta.scan.result <- palm.meta.summary(summary.stats = summary.stat)
```

The following shows 20 coefficient estimates of the microbial features for 5 metabolites.

```{r echo=TRUE}
meta.sum <- lapply(meta.scan.result, function(d){d %>% tibble::column_to_rownames("feature")})
selected.num <- sort(unlist(lapply(meta.sum, function(d){sum(d$qval <= 0.05)})), decreasing = TRUE)
top.cov.name <- names(selected.num)[1:min(5, length(selected.num))]
coef_mat <- do.call(cbind, lapply(meta.sum[top.cov.name], function(d){d[,"coef",drop=FALSE]}))
colnames(coef_mat) <- top.cov.name
coef_mat %>% dplyr::slice(1:20) %>% 
  mutate(across(where(is.numeric), ~ round(.x, digits = 4))) %>% 
  datatable()
```

The following figure displays the features associated with butyrate production with adjusted coefficient $\text{FDR} \leq 0.05E$. Each dot in the scatter plot represents the coefficient estimate, with error bars indicating the standard errors in individual studies. The features with heterogeneity $\text{FDR} \leq 0.1$ are highlighted with a yellow background, this results demonstrate that PALM achieves good harmonization of features across studies, with no heterogeneity detected.

```{r, fig.width=6, fig.height=12, warning=FALSE}
## summarize HMDB0000039 analysis
PALM.df <- meta.scan.result$HMDB0000039 %>% tibble::column_to_rownames("feature") %>% 
  dplyr::filter(qval <= 0.05)

## PALM plots
df.plot <- NULL
for(l in names(otu_data_lst)){
  if(paste0(l,"_effect") %in% colnames(PALM.df)){
    df.plot <- rbind(df.plot, tibble(genus = factor(rownames(PALM.df), levels = rownames(PALM.df)),
                                     Study = factor(rep(l, nrow(PALM.df)), levels = names(otu_data_lst),
                                                    labels = paste0("MTBL", 1:8)),
                                     AA = PALM.df[,paste0(l,"_effect")],
                                     AA.lower = PALM.df[, paste0(l,"_effect")] - PALM.df[, paste0(l,"_stderr")],
                                     AA.upper = PALM.df[, paste0(l,"_effect")] + PALM.df[, paste0(l,"_stderr")]))
  }
}
df.area <- tibble(genus = factor(rownames(PALM.df %>% dplyr::filter(qval.het <= 0.1)), levels= rownames(PALM.df)), AA = 0)

g.PALM <- df.plot %>%
  ggplot(aes(x=genus, y=AA)) +
  geom_errorbar(aes(ymin = AA.lower, ymax = AA.upper, group = Study),
                position = position_dodge(width = 0.6),
                width = 0, linewidth = 0.5) +
  geom_point(pch = 18, aes(color = Study),
             size = 2.5, position = position_dodge(width = 0.6)) +
  ylab("PALM\nCoefficients") + xlab("Genus") +
  geom_hline(aes(yintercept = 0),colour="#990000", linetype="dashed") +
  scale_color_manual(values = c("#F8766D","#CD9600","#7CAE00","#00BE67",
                                "#00BFC4","#00A9FF","#C77CFF","#FF61CC"),
                     breaks = c("MTBL8", "MTBL7", "MTBL6", "MTBL5",
                                "MTBL4", "MTBL3", "MTBL2", "MTBL1")) +
  coord_flip() +
  theme_minimal() +
  theme(axis.title.x = element_text(size = 15),
        axis.title.y = element_blank(),
        axis.ticks = element_blank(),
        panel.grid = element_blank(),
        panel.background = element_rect(fill=NULL, colour='black', linewidth = 1),
        axis.text.y = element_blank(),
        axis.text.x =  element_text(size = 15),
        
        legend.title = element_text(hjust = 0.5, size = 16),
        legend.text = element_text(size = 13),
        legend.position = c(0.2, 0.8),
        legend.direction = "vertical",
        legend.box = "vertical",
        strip.text = element_blank()) +
  guides(color = guide_legend(reverse=T))

if(nrow(df.area) != 0){
  g.PALM <- g.PALM + geom_rect(data = df.area, aes(xmin = as.numeric(genus) - 0.5,
                                                   xmax = as.numeric(genus) + 0.5,
                                                   ymin = -Inf, ymax = Inf),
                               fill = "yellow", alpha = 0.2)
}

g.PALM
```

# Session information

```{r}
sessionInfo()
```

# Reference
1. Zhoujingpeng Wei, Qilin Hong, Guanhua Chen, Tina V. Hartert, Christian Rosas-Salazar, Suman R. Das, and Zheng-Zheng Tang. Fast and reliable association discovery in large-scale microbiome studies and meta-analyses using PALM. Submitted.

2. Wirbel, Jakob et al. Meta-analysis of fecal metagenomes reveals global microbial features that are specific for colorectal cancer. 

3. Muller, E., Algavi, Y.M. & Borenstein, E. The gut microbiome-metabolome dataset collection: a curated resource for integrative meta-analysis. npj Biofilms Microbiomes 8, 79 (2022).
